src/com/amazon/ion/impl/_Private_RecyclingStack.java

```diff
@@ -11,10 +11,16 @@
  * @param <T> the type of elements stored.
  */
 public final class _Private_RecyclingStack<T> implements Iterable<T> {
-
+    private $Iterator stackIterator;
     @Override
     public ListIterator<T> iterator() {
-        return new $Iterator();
+        stackIterator = new $Iterator();
+        return stackIterator;
```

Comment:
I think this will be cleaner if in this method, you
1. Allocate the iterator only if it has not yet been allocated, and
2. Reset the iterator

That way you don't need a public `resetIterator` method, which you call every time you need to retrieve the iterator anyway. Also, you don't need to store a `stackIterator` variable in `IonRawBinaryWriter`, as retrieving it via `containers.iterator()` where you need it will be sufficient. In other words, iterator reuse can be achieved without any changes to IonRawBinaryWriter.

Reply:
Thanks for the suggestions. I experimentally updated method `iterator()` with the following change:
```
public ListIterator<T> iterator() {
        if (stackIterator != null) {
            stackIterator.cursor = _Private_RecyclingStack.this.currentIndex;
        } else {
            stackIterator = new $Iterator();
        }
        return stackIterator;
    }
```
One concern I have for this change is we need to run the `if...else...` condition check every time we call `containers.iterator()`. Then I tried to initialize` $Iterator()` while constructing the `Recycling_Stack` and iterator() method only reset the cursor. This change might cause unnecessary iterator allocation while we do not need to iterate the stack. From the benchmark results the first method (conditional initialize iterator) is more performant. Would there be any other alternative implementations that I wasn't aware of? Thanks


Test Data | conditional allocation | non-conditional allocation
-- | -- | --
log_59155.ion | 496.576 | 501.664
log_194617.ion | 3924.792 | 3965.73

Reply:
The `if/else` is only a problem if it results in a noticeable performance degradation, but it doesn't look like it does. It's surprising that unconditional allocation is slower, but I think the conditional allocation is fine.